contracts/orbital-auction/src/contract.rs

```diff
@@ -0,0 +1,324 @@
+    deps: ExecuteDeps,
+    active_auction: AuctionRound,
+    mock_fill_status: bool,
+) -> NeutronResult<Response<NeutronMsg>> {
+    // if no bids are placed, we finalize the round and prepare for the next one
+    if active_auction.batch.current_bid.is_none() {
+        start_new_round(deps.storage, active_auction)?;
+        process_orderbook(deps.storage)?;
+        return Ok(Response::default());
+    }
+
+    let order_filled = query_order_filling_status(deps.as_ref(), mock_fill_status)?;
+
+    // if the solver succeeded, we do not slash the solver and prepare for the next round.
+    if order_filled {
+        start_new_round(deps.storage, active_auction)?;
+        process_orderbook(deps.storage)?;
+        // TODO: credit the users & solver
+        Ok(Response::default())
+    } else {
+        // if the solver failed to fill the order, we slash the solver, refund the users,
+        // and prepare for the next round.
+        let winning_bid = active_auction.batch.current_bid.clone().unwrap();
+        slash_solver(deps.storage, winning_bid.solver)?;
+
+        // then start the new round
+        start_new_round(deps.storage, active_auction)?;
+        process_orderbook(deps.storage)?;
+
+        // TODO: refund users
+        Ok(Response::default())
+    }
+}
+
+/// moves the user intents from the orderbook queue to the active auction batch
+fn process_orderbook(storage: &mut dyn Storage) -> StdResult<()> {
+    let mut active_auction = ACTIVE_AUCTION.load(storage)?;
+
+    while !ORDERBOOK.is_empty(storage)?
+        && active_auction.batch.remaining_capacity() > Uint128::zero()
+    {
+        if let Some(intent) = ORDERBOOK.pop_front(storage)? {
+            let capacity = active_auction.batch.remaining_capacity();
+            if intent.amount <= capacity {
+                // if it fits, we add it to the batch and update the capacity
+                active_auction.batch.include_order(intent)?;
+            } else {
+                // if it doesn't fit, we check if we could include at least a part of it
+                // and then push the rest back to the orderbook
+                let (fitting, leftover) = intent.split_order(capacity)?;
+                active_auction.batch.include_order(fitting)?;
+                ORDERBOOK.push_front(storage, &leftover)?;
+                break;
+            }
+        }
+    }
+
+    ACTIVE_AUCTION.save(storage, &active_auction)?;
+
+    Ok(())
+}
+
+/// queries the auction config to find the configured bond for this auction
+/// and slashes the solver's bond accordingly.
+fn slash_solver(storage: &mut dyn Storage, solver: Addr) -> StdResult<()> {
+    let auction_config = AUCTION_CONFIG.load(storage)?;
+    let mut posted_bond = POSTED_BONDS.load(storage, solver.to_string())?;
+    posted_bond.amount = posted_bond
+        .amount
+        .checked_sub(auction_config.solver_bond.amount)?;
+
+    POSTED_BONDS.save(storage, solver.to_string(), &posted_bond)?;
+
+    Ok(())
```

Comment:
What is done with the slashed amount? If we are not doing anything with it yet or it's not decided maybe add a TODO here